File: week-08/protoseo/보석 쇼핑.java
import java.util.*;

class Solution {
    Map<String, Integer> gemByCount = new HashMap<>();

    public int[] solution(String[] gems) {
        int[] answer = {-100001, 100001};
        int gemTypeCount = countGemType(gems);
        int l = 0;
        int r = 0;
        while (l <= r && r < gems.length) {
            if (gemByCount.size() < gemTypeCount) {
                String gem = gems[r++];
                gemByCount.put(gem, gemByCount.getOrDefault(gem, 0) + 1);
            } else if (gemByCount.size() == gemTypeCount) {
                if (answer[1] - answer[0] > r - l - 1) {
                    answer[0] = l + 1;
                    answer[1] = r;
                }
                String gem = gems[l++];
                int count = gemByCount.get(gem);
                if (count == 1) {
                    gemByCount.remove(gem);
                } else {
                    gemByCount.put(gem, count - 1);
                }
            }
        }
        while (l <= r && (gemByCount.size() == gemTypeCount)) {
            if (answer[1] - answer[0] > r - l - 1) {
                answer[0] = l + 1;
                answer[1] = r;
            }
            String gem = gems[l++];
            int count = gemByCount.get(gem);
            if (count == 1) {
                gemByCount.remove(gem);
            } else {
                gemByCount.put(gem, count - 1);
            }
        }
        return answer;
    }

    private int countGemType(String[] gems) {
        Set<String> s = new HashSet<>();
        for (String gem : gems) {
            s.add(gem);
        }
        return s.size();
    }
}

File: week-8/dahyen0o/보석-쇼핑.py
def solution(gems):
    counts = {}
    for gem in gems:
        counts[gem] = 0
        
    start, end = 0, 0
    answer = [0, 100_000_000]
    
    counts[gems[0]] += 1
    count = 1
    if count == len(counts) and (answer[1] - answer[0]) > (end - start):
        answer[0] = start
        answer[1] = end
    
    while True:
        if start == end or count < len(counts):
            end += 1
            
            if end == len(gems): break
            counts[gems[end]] += 1
            if counts[gems[end]] == 1:
                count += 1
        else:
            if (answer[1] - answer[0]) > (end - start):
                answer[0] = start
                answer[1] = end
            
            counts[gems[start]] -= 1
            if counts[gems[start]] == 0:
                count -= 1
            start += 1
    
    return list(map(lambda x: x + 1, answer))
